raft.c

```diff
@@ -1339,6 +1411,91 @@ void RaftReqHandleQueue(uv_async_t *handle)
  * Implementation of specific request types.
  */
 
+static void handleTransferLeaderComplete(raft_server_t *raft, raft_transfer_state_e state)
+{
+    raft_node_id_t transfer_id = raft_get_transfer_leader(raft);
+    raft_node_id_t leader = raft_get_leader_id(raft);
+
+    if (!redis_raft.transfer_req) {
+        LOG_ERROR("leader transfer update: but no req to correlate it to!");
+        return;
+    }
+
+    if (!transfer_id) {
+        LOG_ERROR("leader transfer: libraft didn't return a node id we were transferring to");
+        RedisModule_ReplyWithError(redis_raft.transfer_req->ctx, "ERR: no record of a leader transfer");
+        goto exit;
+    }
+
+    switch (state) {
+        case RAFT_STATE_LEADERSHIP_TRANSFER_EXPECTED_LEADER:
+            RedisModule_ReplyWithSimpleString(redis_raft.transfer_req->ctx, "OK");
+            break;
+        case RAFT_STATE_LEADERSHIP_TRANSFER_UNEXPECTED_LEADER:
+            RedisModule_ReplyWithError(redis_raft.transfer_req->ctx, "ERR: transferred to unexpected leader");
+            break;
+        case RAFT_STATE_LEADERSHIP_TRANSFER_TIMEOUT:
+            RedisModule_ReplyWithError(redis_raft.transfer_req->ctx, "ERR: transfer timed out");
+            break;
+        default:
+            RedisModule_ReplyWithError(redis_raft.transfer_req->ctx, "ERR: unknown case");
+            break;
+    }
+
+exit:
+    RaftReqFree(redis_raft.transfer_req);
+    redis_raft.transfer_req = NULL;
+}
+
+/* TODO: move ReplyWith... and RaftReqFree to handleTransferLeaderComplete */
+static void handleTransferLeader(RedisRaftCtx *rr, RaftReq *req)
+{
+    int err;
+
+    if (checkRaftState(rr, req) == RR_ERROR) {
+        goto exit;
+    }
+
+    if ((err = raft_transfer_leader(rr->raft, req->r.node_to_transfer_leader, 0)) != 0) {
+        RedisModuleString *e;
+        switch (err) {
+            case RAFT_ERR_NOT_LEADER:
+                RedisModule_ReplyWithError(req->ctx, "ERR not leader");
+                break;
+            case RAFT_ERR_LEADER_TRANSFER_IN_PROGRESS:
+                RedisModule_ReplyWithError(req->ctx, "ERR transfer already in progress");
+                break;
+            case RAFT_ERR_INVALID_NODEID:
+                e = RedisModule_CreateStringPrintf(req->ctx, "ERR invalid node id: %d", req->r.node_to_transfer_leader);
+                RedisModule_ReplyWithError(req->ctx, RedisModule_StringPtrLen(e, NULL));
+                break;
+            default:
+                e = RedisModule_CreateStringPrintf(req->ctx, "ERR unknown error transferring leader: %d", err);
```

Comment:
I think we need to deallocate these strings. I think you can use local buffer with snprintf instead.

Reply:
But in general, local buffer is used to generate strings to avoid malloc/free in the codebase. If there is no downside, please consider doing something like : 
```c
char buf[128];
...
snprintf("ERR invalid node id: %d", req->r.node_to_transfer_leader);
RedisModule_ReplyWithError(req->ctx, buf);
```

Reply:
ok, understood